docs/elasticsearch-specification.asciidoc

```diff
@@ -585,7 +585,7 @@ Note that this will require restarting the pods.
 
 When an Elasticsearch `Pod` is terminated, its `Endpoint` is removed from the `Service` and the Elasticsearch process is terminated. As these two operations happen in parallel, a race condition exists. If the Elasticsearch process is already shut down, but the `Endpoint` is still a part of the `Service`, any new connection might fail. For more information, see link:https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods[Termination of pods].
 
-To address this issue and guarantee no downtime for new connections, ECK relies on a link:https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/[PreStop lifecycle hook]. This waits to terminate the Elasticsearch process until the `Service` DNS record does not contain the IP of the `Pod`. First, the PreStop lifecycle hook will keep querying DNS for `MAX_WAIT_SECONDS` (defaulting to 20) and then it will wait for `ADDITIONAL_WAIT_SECONDS` (defaulting to 1). Additional wait is used to give clients time to use an IP resolved just before DNS record was updated. The exact behavior is configurable using environment variables, for example:
+To address this issue and guarantee no downtime for new connections, ECK relies on a link:https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/[PreStop lifecycle hook]. This waits to terminate the Elasticsearch process until the `Service` DNS record does not contain the IP of the `Pod`. First, the PreStop lifecycle hook will keep querying DNS for `PRE_STOP_MAX_WAIT_SECONDS` (defaulting to 20) and then it will wait for `PRE_STOP_ADDITIONAL_WAIT_SECONDS` (defaulting to 30). Additional wait is used to give clients time to use an IP resolved just before DNS record was updated. The exact behavior is configurable using environment variables, for example:
```

Comment:
```suggestion
To address this issue and guarantee no downtime for new connections, ECK relies on a link:https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/[PreStop lifecycle hook]. This waits to terminate the Elasticsearch process until the `Service` DNS record does not contain the IP of the `Pod`. First, the PreStop lifecycle hook will keep querying DNS for `PRE_STOP_MAX_WAIT_SECONDS` (defaulting to 20) and then it will wait for `PRE_STOP_ADDITIONAL_WAIT_SECONDS` (defaulting to 30). Additional wait is used to:
1. give clients time to use an IP resolved just before DNS record was updated, and
1. give kube-proxy time to refresh iptables rules on all nodes.
The exact behavior is configurable using environment variables, for example:
```

Also, I think we should mention somehow that additional wait should be at least min-sync-interval, so someone running with higher knows to adjust this and someone running with lower (for example, on GKE) can lower it.

Reply:
> 1. give kube-proxy time to refresh iptables rules on all nodes

Correct me if I'm wrong but should it be "give kube-proxy time **to start refreshing** iptables rules on all nodes" ?
I'm not sure we have any guarantee about when iptables rules will be actually refreshed. IIRC until `ipvs` is more widely used [it can take some time to apply the rules](https://youtu.be/4-pawkiazEg?t=862).

Reply:
The intention here is to allow refresh to start and finish (that's what we are waiting for), ie. we don't care about only starting the refresh. Having said that, you are right that the refresh time is unbounded. Maybe we should say instead that additional time is supposed to be larger than min-sync + the rules refresh?

Reply:
That's correct.
@barkbay @david-kow do you think we should set a higher threshold (more than 30sec) to account for a "slow" iptables refresh?  I could not find any numbers about iptables refresh performance under reasonably normal conditions (ie. not 5k services).

Reply:
I must confess that having fighting some iptables/SDN/DNS problems in production I'm not comfortable with the term "guarantee no downtime for new connections". We are adding some tools and parameters to minimize downtime, which is definitely great, and I'm 👍 with the changes here. My point was more that given the complexity of some deployments (sdn providers, dns caches, ipvs VS iptables) I would not give such guarantee. 
I understand it is a personal feeling though :)

Reply:
> @barkbay @david-kow do you think we should set a higher threshold (more than 30sec) to account for a "slow" iptables refresh?

I think we shouldn't, because:
1) we've are already built in some worst case assumptions (worst possible timing, burst depleted)
2) there is already some "hidden", added time as refresh is on the path happening in parallel to prestop. So when we run prestop, wait for DNS update, es is shutting down, etc. the clock is ticking for iptables refresh.
3) I don't know good value to set it for as it can vary from ~0s to hours as @barkbay noted.

I think I'd document it and assume that if someone has a cluster where routes need 10 mins to refresh for some reason, they can adjust the settings themselves.

Reply:
> I must confess that having fighting some iptables/SDN/DNS problems in production I'm not comfortable with the term "guarantee no downtime for new connections".

The message I tried to convey here is that orchestration done by ECK is not introducing anything that would cause ES downtime (which was the case before). That guarantee doesn't cover parts we are not responsible for (network, other hw, power, etc.), but I'm not sure how to word it differently.

Reply:
> but I'm not sure how to word it differently.

"minimise downtime"?

Reply:
> "minimise downtime"?

At least to me, it implicates the inevitable existence of downtime for all upgrades. But I don't feel that strongly about it :)

Reply:
I agree with all the above 👍 
I kept the 30sec but changed comments and docs to be more explicit about what we're trying to achieve here: minimise unavailability by waiting long enough for a "default" kube-proxy to refresh its rules, which may still take longer in some situations.